### PR TITLE
Add support for custom savers

### PR DESCRIPTION
**Pull request by dmitryuk**
The PR implement possibility to create own Saver, for example, Amazon S3 or StdOut

### CODE REVIEW

**Comment by dmitryuk** [comment]
Why Travis didn't run?

**Comment by glensc** [comment]
Remove **custom** from the methods. let the method just the `setSaver()`.

aside, your own saver you created, it's for debugging purposes? or how are you consuming them? xhgui needs to be able to import from them too?

regarding Travis: travis.org discontinued, need to migrate to travis.com or GitHub actions.

also, never crate pull request from your master branch, as it will pick up any commits you add to your master, which may not be wanted if you want to add other changes to your fork.

**Comment by dmitryuk** [comment]
Changed the method to setSaver.
About the custom saver - in any environment I can export XHProf results to my own storage, for example, Amazon s3 and import to xhgui later using file or API method, so do the job asynchronous.

File: src/Profiler.php [hunk 1]
```diff
@@ -264,6 +274,7 @@ private function getSaver()
         return $this->saveHandler ?: null;
     }
 
+
```
Comment (glensc) [inline src/Profiler.php]: remove superfluous line

File: tests/Saver/NullSaverTest.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?php
```
Comment (glensc) [inline tests/Saver/NullSaverTest.php:1]: wrong filename, class and filename must match

File: tests/Saver/NullSaverTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Xhgui\Profiler\Test\Saver;
+
+use Xhgui\Profiler\Profiler;
+use Xhgui\Profiler\Saver\FileSaver;
+use Xhgui\Profiler\Saver\SaverInterface;
+use Xhgui\Profiler\Test\TestCase;
+
+/**
+ * @requires extension json
```
Comment (glensc) [inline tests/Saver/NullSaverTest.php]: bogus dependency. remove

File: tests/Saver/NullSaverTest.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            $profiler->stop();
+            $this->markTestIncomplete('Custom saver not executed');
+        } catch (\Exception $e) {
+            $this->assertEquals('CustomSaver executed', $e->getMessage());
+        }
+    }
+}
+
+
+class NullSaver implements SaverInterface
```
Comment (glensc) [inline tests/Saver/NullSaverTest.php]: one class per file, see psr-1/psr-2

File: tests/Saver/NullSaverTest.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+{
+    public function isSupported()
+    {
+        return true;
+    }
+
+    public function save(array $data)
+    {
+        throw new \Exception('CustomSaver executed');
```
Comment (glensc) [inline tests/Saver/NullSaverTest.php]: name mismatch

File: README.md [hunk 6]
```diff
@@ -252,6 +253,31 @@ Example config:
     ),
 ```
 
+## Custom Saver
+You may create custom profiler saver by implementing SaverInterface.
```
Comment (glensc) [inline README.md]:
```suggestion
You can create your own profile saver by implementing `SaverInterface` and calling `setSaver()`.
```

File: tests/Saver/NullSaverTest.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    public function setCustomSaver()
+    {
+        $saver = new NullSaver();
+        $profiler = new Profiler(array());
+        $profiler->setSaver($saver);
+        $profiler->start();
+        try {
+            $profiler->stop();
+            $this->markTestIncomplete('Custom saver not executed');
+        } catch (\Exception $e) {
```
Comment (glensc) [inline tests/Saver/NullSaverTest.php]: Import exception (use Exception)

**Review by dmitryuk — COMMENTED** [review]
All review comments fixed.

**Comment by markstory** [comment]
How does custom saver data get back into xhgui? Is that left up to the implementers of custom savers?

**Comment by dmitryuk** [comment]
> About the custom saver - in any environment I can export XHProf results to my own storage, for example, Amazon s3 and import to xhgui later using file or API method, so do the job asynchronous.

This way for example.

File: src/Profiler.php [hunk 8]
```diff
@@ -298,4 +308,4 @@ private function getDefaultConfig()
             'profiler.replace_url' => null,
         );
     }
-}
```
Comment (glensc) [inline src/Profiler.php]: remove this change

File: tests/Resources/NullSaver.php [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    public function isSupported()
+    {
+        return true;
+    }
+
+    public function save(array $data)
+    {
+        throw new Exception('NullSaver executed');
+    }
+}
```
Comment (glensc) [inline tests/Resources/NullSaver.php]: missing newline

**Comment by glensc** [comment]
Tests were not actually passing: https://github.com/perftools/php-profiler/pull/71

**Pull request merged**